packages/repository/src/repositories/relation.factory.ts

```diff
@@ -35,18 +37,19 @@ export interface HasManyDefinition extends RelationDefinitionBase {
  * relation attached to a datasource.
  *
  */
-export function hasManyRepositoryFactory<SourceID, T extends Entity, ID>(
-  sourceModelId: SourceID,
-  relationMetadata: HasManyDefinition,
-  targetRepository: EntityCrudRepository<T, ID>,
-): HasManyEntityCrudRepository<T> {
-  switch (relationMetadata.type) {
-    case RelationType.hasMany:
-      const fkConstraint = {[relationMetadata.keyTo]: sourceModelId};
-
-      return new DefaultHasManyEntityCrudRepository<
-        T,
-        EntityCrudRepository<T, ID>
-      >(targetRepository, fkConstraint);
-  }
+export function createHasManyRepositoryFactory<
+  S extends Entity,
+  T extends Entity
+>(
+  relationMeta: HasManyDefinition,
+  targetRepo: EntityCrudRepository<T, typeof Entity.prototype.id>,
+) {
+  return function(constraint: Partial<S>) {
+    const fkName = relationMeta.keyTo;
+    const fkValue = constraint[relationMeta.keyFrom];
+    return new DefaultHasManyEntityCrudRepository<
+      T,
+      EntityCrudRepository<T, typeof Entity.prototype.id>
+    >(targetRepo, {[fkName]: fkValue});
```

Comment:
Maybe instead of `ID extends keyof T`, we can allow arbitrary `ID`. In the future, this `ID` can be specified as an object type containing multiple properties building the multi-column primary key. (For example, a GitHub issue can have primary key type as`ID={user:string, repo: string, id: number}`.)